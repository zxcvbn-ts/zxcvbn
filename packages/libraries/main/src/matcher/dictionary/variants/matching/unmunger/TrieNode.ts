export default class TrieNode {
  constructor(public parents: string[] = []) {}

  // eslint-disable-next-line no-use-before-define
  children: Map<string, TrieNode> = new Map()

  subs?: string[]

  addSub(key: string, ...subs: string[]): TrieNode {
    const firstChar = key.charAt(0)
    if (!this.children.has(firstChar)) {
      this.children.set(firstChar, new TrieNode([...this.parents, firstChar]))
    }
    let cur = this.children.get(firstChar)!
    for (let i = 1; i < key.length; i += 1) {
      const c = key.charAt(i)
      if (!cur.hasChild(c)) {
        cur.addChild(c)
      }
      cur = cur.getChild(c)!
    }
    cur.subs = (cur.subs || []).concat(subs)
    return this
  }

  getChild(child: string): TrieNode | undefined {
    return this.children.get(child)
  }

  isTerminal(): boolean {
    return !!this.subs
  }

  addChild(child: string): void {
    if (!this.hasChild(child)) {
      this.children.set(child, new TrieNode([...this.parents, child]))
    }
  }

  hasChild(child: string): boolean {
    return this.children.has(child)
  }
}
