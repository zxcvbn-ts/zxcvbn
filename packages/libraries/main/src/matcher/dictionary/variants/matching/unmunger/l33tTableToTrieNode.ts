import { OptionsL33tTable } from '../../../../../types'
import TrieNode from './TrieNode'

export default (l33tTable: OptionsL33tTable, triNode: TrieNode) => {
  Object.entries(l33tTable).forEach(([letter, substitutions]) => {
    substitutions.forEach((substitution) => {
      triNode.addSub(substitution, letter)
    })
  })
  return triNode
}
