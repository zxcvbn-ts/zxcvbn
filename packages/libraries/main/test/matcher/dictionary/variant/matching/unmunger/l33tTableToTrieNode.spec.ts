import l33tTableToTrieNode from '../../../../../../src/matcher/dictionary/variants/matching/unmunger/l33tTableToTrieNode'
import TrieNode from '../../../../../../src/matcher/dictionary/variants/matching/unmunger/TrieNode'

describe('l33t table to trie node', () => {
  it('should convert correctly', () => {
    expect(
      l33tTableToTrieNode(
        {
          a: ['@', '4'],
          u: ['|_|'],
          m: ['^^', 'nn', '2n', '/\\\\/\\\\'],
        },
        new TrieNode(),
      ),
    ).toMatchSnapshot()
  })
})
