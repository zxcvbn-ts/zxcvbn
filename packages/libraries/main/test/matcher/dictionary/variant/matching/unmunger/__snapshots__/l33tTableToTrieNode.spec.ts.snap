// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`l33t table to trie node should convert correctly 1`] = `
TrieNode {
  "children": Map {
    "@" => TrieNode {
      "children": Map {},
      "parents": [
        "@",
      ],
      "subs": [
        "a",
      ],
    },
    "4" => TrieNode {
      "children": Map {},
      "parents": [
        "4",
      ],
      "subs": [
        "a",
      ],
    },
    "|" => TrieNode {
      "children": Map {
        "_" => TrieNode {
          "children": Map {
            "|" => TrieNode {
              "children": Map {},
              "parents": [
                "|",
                "_",
                "|",
              ],
              "subs": [
                "u",
              ],
            },
          },
          "parents": [
            "|",
            "_",
          ],
        },
      },
      "parents": [
        "|",
      ],
    },
    "^" => TrieNode {
      "children": Map {
        "^" => TrieNode {
          "children": Map {},
          "parents": [
            "^",
            "^",
          ],
          "subs": [
            "m",
          ],
        },
      },
      "parents": [
        "^",
      ],
    },
    "n" => TrieNode {
      "children": Map {
        "n" => TrieNode {
          "children": Map {},
          "parents": [
            "n",
            "n",
          ],
          "subs": [
            "m",
          ],
        },
      },
      "parents": [
        "n",
      ],
    },
    "2" => TrieNode {
      "children": Map {
        "n" => TrieNode {
          "children": Map {},
          "parents": [
            "2",
            "n",
          ],
          "subs": [
            "m",
          ],
        },
      },
      "parents": [
        "2",
      ],
    },
    "/" => TrieNode {
      "children": Map {
        "\\" => TrieNode {
          "children": Map {
            "\\" => TrieNode {
              "children": Map {
                "/" => TrieNode {
                  "children": Map {
                    "\\" => TrieNode {
                      "children": Map {
                        "\\" => TrieNode {
                          "children": Map {},
                          "parents": [
                            "/",
                            "\\",
                            "\\",
                            "/",
                            "\\",
                            "\\",
                          ],
                          "subs": [
                            "m",
                          ],
                        },
                      },
                      "parents": [
                        "/",
                        "\\",
                        "\\",
                        "/",
                        "\\",
                      ],
                    },
                  },
                  "parents": [
                    "/",
                    "\\",
                    "\\",
                    "/",
                  ],
                },
              },
              "parents": [
                "/",
                "\\",
                "\\",
              ],
            },
          },
          "parents": [
            "/",
            "\\",
          ],
        },
      },
      "parents": [
        "/",
      ],
    },
  },
  "parents": [],
}
`;
